File: Unique rows in boolean matrix - GFG/unique-rows-in-boolean-matrix.py
#User function Template for python3

def uniqueRow(row, col, matrix):
    #complete the function
    mydict = {}
    ans = []
    i = 0
    while i < len(matrix):
        row = matrix[i:i+col]
        if tuple(row) not in mydict:
            mydict[tuple(row)] = 0
            ans.append(row)
        i+=col
    return ans

#{ 
 # Driver Code Starts
#Initial Template for Python 3

def main():
    testcase = int(input())
    while(testcase):
        s = input().split()
        row = int(s[0])
        col = int(s[1])
        matrix = input().split()
        
        a = uniqueRow(row, col, matrix)
        
        for row in a:
            for value in row:
                print(value,end = " ")
            print("$",end = "")
        print()
        testcase -= 1

if __name__ == "__main__":
    main()
# } Driver Code Ends

File: Unique rows in boolean matrix - GFG/unique-rows-in-boolean-matrix.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;
#define MAX 1000

vector<vector<int>> uniqueRow(int M[MAX][MAX],int row,int col);

int main()
{
    int T;
    cin>>T;
    while(T--)
    {
    	int n,m;
    	cin>>n>>m;
    	int a[MAX][MAX];
    	for(int i=0;i<n;i++)
    	for(int j=0;j<m;j++)
    	cin>>a[i][j];
    	vector<vector<int>> vec = uniqueRow(a,n,m);
    	for(int i = 0;i<vec.size();i++){
    	    for(int x : vec[i]){
    	        cout<<x<<" ";
    	    }
    	    cout<<"$";
    	}
    	cout<<endl;
    }
}

// } Driver Code Ends


/*You are required to complete this function*/
struct TireNode{
    bool isend;
    struct TireNode* child[2];
    TireNode(){
        isend = false;
        child[0] = child[1] = nullptr;
    }
};
vector<vector<int>> uniqueRow(int M[MAX][MAX],int row,int col)
{
//Your code here
    vector<vector<int>> ans;
    TireNode *root = new TireNode();
    for(int i = 0; i < row; i++){
        TireNode *curr = root;
        for(int j = 0; j < col; j++){
            if(!curr -> child[M[i][j]])
                curr -> child[M[i][j]] = new TireNode();
            curr = curr -> child[M[i][j]];
        }
        if(!curr -> isend){
            vector<int> temp(col);
            copy(M[i],M[i] + col, temp.begin());
            ans.push_back(temp);
            curr -> isend = true;
        }
    }
    return ans;
}